cmd/docker/docker.go

```diff
@@ -187,16 +191,53 @@ func setValidateArgs(dockerCli command.Cli, cmd *cobra.Command) {
 	})
 }
 
+func setupPluginSocket() (*net.UnixListener, error) {
+	uuid := uuid.Generate().String()
+	pluginCloseSocketAddr := net.UnixAddr{Name: "@docker_cli_" + uuid, Net: "unix"}
+	return net.ListenUnix("unix", &pluginCloseSocketAddr)
+}
+
 func tryPluginRun(dockerCli command.Cli, cmd *cobra.Command, subcommand string, envs []string) error {
 	plugincmd, err := pluginmanager.PluginRunCommand(dockerCli, subcommand, cmd)
 	if err != nil {
 		return err
 	}
 	plugincmd.Env = append(envs, plugincmd.Env...)
 
+	var conn *net.UnixConn
+	listener, err := setupPluginSocket()
+	if err == nil {
+		defer listener.Close()
+		plugincmd.Env = append(plugincmd.Env, plugin.CliPluginCloseSocketEnvKey+"="+listener.Addr().String())
+
+		go func() {
+			// ignore error here, if we failed to accept a connection,
+			// conn is nil and we fallback to previous behavior
+			conn, _ = listener.AcceptUnix()
+		}()
+	}
+
+	signals := make(chan os.Signal, 2048)
```

Comment:
```suggestion
	signals := make(chan os.Signal, exitLimit)
```
There's no need to buffer any more signals as the process will have exited after processing at most three of them.